drivers/espi/espi_handlers.c

```diff
@@ -0,0 +1,146 @@
+/*
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/espi.h>
+#include <syscall_handler.h>
+
+
+static inline int z_vrfy_espi_config(struct device *dev, struct espi_cfg *cfg)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, config));
+       Z_OOPS(Z_SYSCALL_MEMORY_READ(cfg, sizeof(*cfg)));
+
+       return z_impl_espi_config(dev, cfg);
+}
+#include <syscalls/espi_config_mrsh.c>
+
+static inline bool z_vrfy_espi_get_channel_status(struct device *dev,
+						  enum espi_channel ch)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, get_channel_status));
+
+       return z_impl_espi_get_channel_status(dev, ch);
+}
+#include <syscalls/espi_get_channel_status_mrsh.c>
+
+static inline int z_vrfy_espi_read_lpc_request(struct device *dev,
+					       enum lpc_peripheral_opcode op,
+					       u32_t *data)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, read_lpc_request));
+       Z_OOPS(Z_SYSCALL_MEMORY_WRITE(data, sizeof(*data)));
+
+       return z_impl_espi_read_lpc_request(dev, op, data);
+}
+#include <syscalls/espi_read_lpc_request_mrsh.c>
+
+static inline int z_vrfy_espi_write_lpc_request(struct device *dev,
+						enum lpc_peripheral_opcode op,
+						u32_t *data)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, write_lpc_request));
+       Z_OOPS(Z_SYSCALL_MEMORY_READ(data, sizeof(*data)));
+
+       return z_impl_espi_write_lpc_request(dev, op, data);
+}
+#include <syscalls/espi_write_lpc_request_mrsh.c>
+
+static inline int z_vrfy_espi_send_vwire(struct device *dev,
+					 enum espi_vwire_signal signal,
+					 u8_t level)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, send_vwire));
+
+       return z_impl_espi_send_vwire(dev, signal, level);
+}
+#include <syscalls/espi_send_vwire_mrsh.c>
+
+static inline int z_vrfy_espi_receive_vwire(struct device *dev,
+					    enum espi_vwire_signal signal,
+					    u8_t *level)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, receive_vwire));
+       Z_OOPS(Z_SYSCALL_MEMORY_WRITE(level, sizeof(*level)));
+
+       return z_impl_espi_receive_vwire(dev, signal, level);
+}
+#include <syscalls/espi_receive_vwire_mrsh.c>
+
+static inline int z_vrfy_espi_read_request(struct device *dev,
+					   struct espi_request_packet *req)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, read_request));
+       Z_OOPS(Z_SYSCALL_MEMORY_WRITE(req, sizeof(*req)));
+       Z_OOPS(Z_SYSCALL_MEMORY_WRITE(req->data, req->len));
+
+       return z_impl_espi_read_request(dev, req);
+}
+#include <syscalls/espi_read_request_mrsh.c>
+
+static inline int z_vrfy_espi_write_request(struct device *dev,
+					    struct espi_request_packet *req)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, read_request));
+       Z_OOPS(Z_SYSCALL_MEMORY_READ(req, sizeof(*req)));
+       Z_OOPS(Z_SYSCALL_MEMORY_READ(req->data, req->len));
```

Comment:
same as above

Reply:
there's more, a lot of these need to be modified.

Reply:
Oww indeed, I was sure I was missing something. Thanks for pointing it.

Reply:
Btw that is one thing that I was thinking about. We really need to copy_to/from in these structs with pointers fields since we don't want end up writing/reading invalid memory due TOCTOU.
But all pointers are potentially problematic, we will of course ensure that the calling thread has right permission to that address but the pointer contents can be successive to TOCTOU right ?

I *think* Linux always copy_to/from pointers in syscalls

Reply:
> I think Linux always copy_to/from pointers in syscalls

Yes I believe we need to do the same.